activerecord/lib/active_record/attribute_methods/primary_key.rb

```diff
@@ -21,6 +21,12 @@ def id
         @primary_key.map { |pk| _read_attribute(pk) }
       end
 
+      def primary_key_values_present? # :nodoc:
+        return id.size > 0 && id.all? if self.class.composite_primary_key?
+
+        !!id
```

Comment:
We can't return `nil` as a falsey value here as it will lead to `==` return `nil` instead of a boolean if the chain of `&&` checks breaks on `primary_key_values_present?`